Uniquely identify tests

MisanthropicBit opened the pull request:
With the current implementation, running the test in either `suite 1` or `suite 2` will run the other test as well because the pattern that is constructed and passed to mocha doesn't account for the suite name.

```javascript
describe('top-level suite', () => {
  describe('suite 1', () => {
    it('tests a test', async () => {
      console.log("first test")
    })
  })
  describe('suite 2', () => {
    it('tests a test', async () => {
      console.log("second test")
    })
  })
})
```

These changes follow the code in [neotest-jest](https://github.com/nvim-neotest/neotest-jest/blob/main/lua/neotest-jest/init.lua#L381-L395) with some modification. Hopefully the code and test are straight-forward.

Failing test is related to neotest's treesitter module.

MisanthropicBit commented on lua/neotest-mocha/init.lua:
```diff
@@ -175,12 +175,11 @@ function Adapter.build_spec(args)
+  if pos.type == "test" or pos.type == "namespace" then
+    local testName = string.sub(pos.id, string.find(pos.id, "::") + 2)
+    testName, _ = string.gsub(testName, "::", " ")
+    testNamePattern = "'^" .. util.escape_test_pattern(testName)
```
This logic has changed to always include a `'^'` since we want to match the specific full position id (without filename).

adrigzr commented:
Hi @MisanthropicBit, thanks for your PR! Could you take a look at the failing jobs? 🙏

MisanthropicBit commented:
I had some time to look at the issue which stems from [this neotest PR](https://github.com/nvim-neotest/neotest/pull/298) which changes [the failing line](https://github.com/nvim-neotest/neotest/blob/master/lua/neotest/lib/treesitter/init.lua#L117). It is listed as a breaking change that requires at least neovim 0.9.0 (as seen from [the release notes](https://github.com/nvim-neotest/neotest/releases/tag/v4.0.0)) which is why the test fails with 0.7.0.

Options would be:

* To conditionally install an earlier version of neotest for the test on 0.7.0.
* Remove the 0.7.0 test from the workflow if you wish for it to run against the latest neotest version.

This isn't technically a blocker for this PR since it has nothing to do with my changes, just that neotest got updated meanwhile, but we might as well fix it now 🙂

MisanthropicBit commented:
I've removed the 0.7.0 test from the workflow and added a note about requiring at least neotest 4.0.0.

Now the tests against nightly fail, likely because something happened with treesitter upstream. I've verified that the query works with `:TSPlaygroundToggle` so I'll try to see what changed (if this is intentional) or if it's a bug in nightly when I have some more time.

MisanthropicBit commented:
@adrigzr So I think I figured out the issue:

* Two weeks ago, the javascript tree-sitter parser [updated the `function` keyword](https://github.com/tree-sitter/tree-sitter-javascript/commit/84c57eeb0eb28046e72f08d77b4c00efc5819a24) so it didn't collide with the `function` (parser) literal with v0.20.2.
* This change was then [included](https://github.com/nvim-treesitter/nvim-treesitter/commit/6da205148fb43eb4fa70888fd6ac46a32f1fc3f2) in nightly nvim-treesitter two weeks ago as well causing the tests here to fail.

I've updated the `discover_positions` function to use `function_expression` instead which solves the issue but this is an issue for users that don't have nvim-treesitter >= 0.9.2 so perhaps the query should be conditional on the nvim-treesitter version?

adrigzr commented:
Hello @MisanthropicBit, the change looks good to me. I'm merging it with the main branch. Thanks a lot!
